### ISSUE TITLE
[FEA] Rename `cudf::null_order` enum to `cudf::null_comparison`

### ISSUE THREAD

**Issue by ttnghia** [rapidsai/cudf#8269]
Currently, we have the `null_order` enum class in `cudf/types.hpp`:
```
/**
 * @brief Indicates how null values compare against all other values.
 */
enum class null_order : bool {
  AFTER,  ///< NULL values ordered *after* all other values
  BEFORE  ///< NULL values ordered *before* all other values
};
```

We use this enum to specify how a value is compared to a null, and it is used a lot. However, the value name (`BEFORE` and `AFTER`) only make sense when comparing (and sorting) in ascending order. For descending order, those names are very confusing.

For example, when we want to sort an array `[1, 2, null]` in descending order, and specify `null_order::AFTER`, the correct result will be `[null, 2, 1]`. As we can see, the null element is placed at the beginning of the array, not "AFTER". Its meaning here is  "GREATER", not "AFTER". Similar situation when we sort the array in descending order with `null_order::BEFORE`---the result will be `[2, 1, null]`. Its meaning is "SMALLER" instead. So, the current enum does not have a good name.

**Request:** Rename the `null_order` enum into this (or similar, better name):
```
enum class null_comparison : bool {
  SMALLER,
  GREATER
};
```

**Comment by harrism** [comment]
[Warning: grammatical pedantry]

I agree, however since nulls are always placed either at the very beginning or the very end, I would say it should be LEAST and GREATEST. In grammatical terms, superlative rather than comparative. Note it should be either smallest/largest or least/greatest, not smallest/greatest or least/largest.

```
enum class null_comparison : bool {
  LEAST,
  GREATEST
};
```

**Comment by jrhemstad** [comment]
> Its meaning here is "GREATER", not "AFTER". Similar situation when we sort the array in descending order with `null_order::BEFORE`---the result will be `[2, 1, null]`. Its meaning is "SMALLER" instead. So, the current enum does not have a good name.

What does it mean for one value to be "GREATER" than another?

**Comment by ttnghia** [comment]
> What does it mean for one value to be "GREATER" than another?

"GREATER" and "LESS" are the relations that we used when sorting elements. Base on those relations, we can determine which element should stay before the others, in a consistent way. If we use "BEFORE" null_order, during sorting we still need to determine whether a null is "GREATER" than a valid element. Thus, having both "GREATER" relation and "BEFORE" null_order will make things complicated and error-prone: you need to decide when a null is "GREATER" a valid element based on the sorting order and the null_order.

**Comment by jrhemstad** [comment]
> Base on those relations, we can determine which element should stay **before** the others, in a consistent way

Your description reveals precisely the reason that the current naming is the way that it is.

The formal, general definition of the inequality operators is that if `a < b`, then `a` is _ordered before_ `b`. Likewise, if `a > b`, `a` is _ordered after_ `b`. 

Whether you are doing an ascending or descending sort is irrelevant. The _ordered before_ and _ordered after_ relationship remains the same. 

Mathematically speaking, a null element has no value. Therefore it doesn't really make sense to describe it as "greatest" or "least" because the absence of a value cannot have extent. The only thing that makes sense is to discuss its ordering in relation to real values, hence why the naming is the way it is.

**Comment by ttnghia** [comment]
> if `a > b`, `a` is _ordered after_ `b`.

Yeah, that's logical, but it does not make sense to me if the given null order is "BEFORE" but after sorting nulls are placed "AFTER" everything else. I believe that we can do better to eliminate such confusion.

**Comment by GregoryKimball** [comment]
I agree that the name `BEFORE` is confusing in the case of descending order sorting. 
* One option could be sort the valid values ascending or descending, followed by placing the nulls at the beginning or end. This way `null_order::BEFORE` would result in nulls ordered in the beginning for both ascending and descending sorted values. However, this would be a breaking change that would introduce more confusion that it would resolve. 
* I agree that null elements cannot be described as `LEAST` or `GREATEST`, so the specific proposed wording would not be appropriate.

For now I'll close the issue until there is a push to revisit libcudf sorting conventions.

**Comment by bdice** [comment]
This naming question came up again in conversation with @divyegala in #12558 because the choice of names and behavior both differ from SQL (and Spark) which use `NULLS FIRST` or `NULLS LAST`. Those SQL/Spark options only affect top level nulls (it seems that lists containing nulls are always sorted such that `[null] < [1]` regardless of that option).

I understand the argument for the terminology _ordered before_. That is a sensible argument based in mathematical terms, but it conflicted with my expectations when used in conjunction with a descending sort. My intuition was that `AFTER` would mean "later in the result list" regardless of ascending or descending sorting. If it were named `ORDERED_AFTER` (or `LESS/GREATER` or `LEAST/GREATEST`), I think I would have realized this was an "ordering" definition of the comparator (and thus the ordering would be flipped when using a descending sort). Something about the phrase "ordered before" is clearer than "before" to me, in terms of how it interacts with sort ordering.

**Comment by GregoryKimball** [comment]
After seeing this line in [python/cudf/cudf/_lib/sort.pyx](https://github.com/rapidsai/cudf/blob/5f6d9d41c4a685632eb2dc8f8f1b163d18d2ba30/python/cudf/cudf/_lib/sort.pyx#L132), I've decided to reopen this issue.

```
        if asc:
            column_order.push_back(order.ASCENDING)
        else:
            column_order.push_back(order.DESCENDING)

        if asc ^ (na_position == "first"):
            null_precedence.push_back(null_order.AFTER)
        else:
            null_precedence.push_back(null_order.BEFORE)
```

**Comment by jrhemstad** [comment]
I think people get confused by this because they think about the position of null elements in the overall set of elements. That makes sense when your perspective comes from Pandas with the `na_position` argument.

However, I never had that context and that is never what `null_order` was intended to convey. I wasn't thinking about the position in the overall set because I didn't know at the time if it would be sufficient to put null elements as only "first" or "last", e.g., I didn't know if we'd want to be able to have NaNs before nulls, etc.  I was writing a comparator, so I naturally was focused on the comparison between two elements `a` and `b`. Formally, `a < b` means `a` is _ordered before_ `b` and `b > a` means `b` is _ordered after_ `a`.

When comparing a null element `N` with any real valued element `x` I wanted to control the result of `N < x`.  This result is completely independent of whether or not the result is ascending or descending. A comparator doesn't care about ascending/descending, it only cares about establishing the `a < b` ordered-before relationship.

In contrast, if we only had "first" and "last" options, then this would require inserting extra logic into the comparator as the result of `N < x` would now depend on whether or not the result is ascending/descending. 

@GregoryKimball this is exactly the logic I wanted to avoid putting into the comparator:
```
        if asc ^ (na_position == "first"):
            null_precedence.push_back(null_order.AFTER)
        else:
            null_precedence.push_back(null_order.BEFORE)
```

TL;DR: I've no problem with adding a nulls_first/nulls_last enum to the public facing API for ordering functions, but I would advocate against removing/changing the `null_order::BEFORE/AFTER` in the context of a comparator. One would just need to map the `nulls::first/last` to nulls::before/after` values depending on if it is ascending or descending, though it looks like this is already being done at the Python level.

**Comment by bdice** [comment]
@jrhemstad Totally agree this approach is sensible for the internals of a comparator. On the downside, it makes the libcudf API a tiny bit clunky because the change in conventions based on ascending/descending is needed for each framework libcudf serves: Spark/SQL and Pandas. It could be argued that we should handle the first/last option in libcudf's user API and then translate internally as needed, but I don't feel strongly about that.

More importantly, however, there are different requirements for element comparison (imposed by both Spark/SQL and Pandas) based on whether they're in a list or not. The options for Spark/SQL `NULLS FIRST/LAST` and pandas `na_position : {"first", "last}` only control top-level nulls. In both Spark/SQL and pandas, nulls in lists are always handled such that `[null] < [1]` regardless of `NULLS FIRST/LAST` or `na_position`. (That is, it's ordered `[null], [1]` when sorted ascending and `[1], [null]` when sorted descending.) We need to control the sorting of top level nulls differently than nulls in lists, so libcudf's element comparator might need to know the depth of nesting (or have multiple element comparators with different null behaviors). At that point, we are faced with a fairly significant refactoring that might warrant a name change or other way to indicate that the null ordering only affects top level nulls. I don't think we have a compelling requirement to keep support for null ordering in lists (to permit `[1] < [null]`), so I'd like to drop that from the comparator entirely.

**Comment by GregoryKimball** [comment]
Thanks @jrhemstad and @bdice for sharing these excellent points.

> A comparator doesn't care about ascending/descending, it only cares about establishing the `a < b` ordered-before relationship
Agreed. This is a great reason to leave `null_order` implementation as it is.

> I've no problem with replacing the `null_order` enum with a `null_position` enum in the public facing API 
I like this proposal for making the API more familiar to database and dataframe users. However, the scale of the refactoring is too big to justify at this time, with >100 references across many algorithms. We would also have to decide where the detail API would use continue to use `null_order` or switch to `null_position`.
```
cpp/include/cudf/merge.hpp
cpp/include/cudf/aggregation.hpp
cpp/include/cudf/groupby.hpp
cpp/include/cudf/lists/sorting.hpp
cpp/include/cudf/quantiles.hpp
cpp/include/cudf/search.hpp
cpp/include/cudf/sorting.hpp
cpp/include/cudf/table/row_operators.hpp
```

> In both Spark/SQL and pandas, nulls in lists are always handled such that `[null] < [1]` regardless of `NULLS FIRST/LAST` or `na_position`. 
Thanks @bdice for including list sorting in the discussion. This comment extends the discussion to list lexicographic comparison in #12538. It seems to me that the `null_order` concept should only apply to top level nulls, and we will need to use a different comparator for nested nulls.

> I don't think we have a compelling requirement to keep support for null ordering in lists (to permit `[1] < [null]`), so I'd like to drop that from the comparator entirely.
It sounds like we agree that our first implementation will support top level null order, and use a default nested null order of "before". If we found a user that needed more control we could consider a `nested_null_order` enum or a larger refactor at that point. 

* I'll close this issue and we can continue discussion of nested null order in #12538. 
* I'm glad we discussed the possibility of a `null_position` parameter in the public API for ordering functions. It is a proposal that would address the confusion for Spark and pandas developers, with the downside being significant work for no performance or functionality gain.
